draft-ietf-tls-esni.md

```diff
@@ -723,38 +716,62 @@ error code.
-the new EncodedClientHelloInner value as a second message with the previous
-HPKE context:
+As described in {{server-hrr}}, if the server has accepted ECH but needs to send
+an HRR then it will be carried in an encrypted HRR value in an
+"encrypted_client_hello" extension inside an outer HRR. If the server sends a
+HelloRetryRequest (HRR) in response to the ClientHello, the client first checks
+for the presence of the "encrypted_client_hello" extension. The presence of this
+extension is an indication that the client-facing server accepted ECH and that
+the backend server sent an HRR. If absent, then the client presumes rejection
+and proceeds as if the message was constructed by the client-facing server.
+
+The client must therefore distinguish between two HRR messages; one constructed
+by the backend server and another constructed by the client-facing server. In
+the remainder of this document, the HRR constructed by the backend server is
+called the HelloRetryRequestInner message; and the HRR constructed by the
+client-facing server is called the HelloRetryRequestOuter message.
+
+#### Accepted ECH {#client-hrr-accepted-ech}
+
+As described in {{sever-hrr}, the HelloRetryRequestInner handshake message is
+encrypted by the backend server and included as the payload of the
+"encrypted_client_hello" extension of the HelloRetryRequestOuter. The client
+decrypts the payload using the HPKE context it shares with the client-facing
+server. It begins by exporting a key and nonce for AEAD decryption:
+
+~~~
+    hrr_inner_key = context.Export("hrr inner key", Nk)
+    hrr_inner_nonce = context.Export("hrr inner nonce", Nn)
+~~~
+
+Here `Nk` and `Nn` denote, respectively, the key and nonce length of the AEAD
+algorithm associated with `context`. Let `encrypted_hrr_inner` denote the
+extension payload. The client computes
+
+~~~
+    HelloRetryRequestInner = Open(hrr_inner_key,
+                                  hrr_inner_nonce,
+                                  HelloRetryRequestOuterAAD
+                                  encrypted_hrr_inner)
+~~~
+
+where HelloRetryRequestOuterAAD is computed by serializing
+HelloRetryRequestOuter with the "encrypted_client_hello" extension removed.
+
+If decryption fails, then the client MUST abort the connection with an
+"illegal_parameter" alert.
+
+The second ClientHelloOuter is identical to the first, except that the
+extensions field contains only the "encrypted_client_hello" extension, which
+contains the second ClientHelloInner, and any extension mandated by {{RFC8446}}.
+This includes [[TODO: Enumerate the extensions that must appear here. This list
+includes "supported_versions", for example.]] The client uses
+HelloRetryRequestInner to construct ClientHelloInner per the rules of
+{{RFC8446}}. It then encodes the second ClientHelloInner as in
+{{encoding-inner}}, except that it does not incorporate any extensions from the
+ClientHelloOuter (i.e., the "ech_outer_extensions" extension is not used). It
```

Comment:
> (i.e., the "ech_outer_extensions" extension is not used)

Why is this?

Reply:
The idea here is that there's no need to incorporate outer extensions because since the inner handshake was taken. However, it occurs to me that this assumption might not be correct, depending on how the "TODO" above is resolved. I wonder, are there any extensions that MUST appear in CHOuter (according to the HRR rules prescribed by 8446), but which one would want to incorporate here? "supported_versions" is the only extension I know of that MUST appear in the CHOuter, but there's no reason to incorporate it.